File: npbloom-web/src/content/positioned/positionedEntityHelpers.ts
import { associateWith, filterEntries, isEmpty, mapValues } from '../../util/objTransforms';
import { Id, IdMap, StringSlice } from '../types';
import slicesOverlap from '../slicesOverlap';
import { PositionedNode, PositionedTree } from './types';

/**
 * Returns nodes matching the given predicate in the given tree.
 */
export const filterPositionedNodesInTree =
  (predicate: (node: PositionedNode) => boolean) =>
  (tree: PositionedTree): IdMap<PositionedNode> =>
    filterEntries(tree.nodes, ([_, node]) => predicate(node));

/**
 * Returns nodes with the given IDs in the given tree.
 */
export const filterPositionedNodesInTreeById =
  (nodeIds: Id[]) =>
  (tree: PositionedTree): IdMap<PositionedNode> =>
    filterEntries(tree.nodes, ([nodeId, _]) => nodeIds.includes(nodeId));

/**
 * Returns whether the node with the given ID is a top-level node in the given node ID map.
 */
export const isTopLevel = (nodes: IdMap<PositionedNode>) => (nodeId: Id) =>
  isEmpty(filterEntries(nodes, ([_, node]) => 'children' in node && node.children.includes(nodeId)));

/**
 * Returns an ID map consisting of the top-level nodes in the given tree.
 */
export const getTopLevelPositionedNodes = (tree: PositionedTree): IdMap<PositionedNode> =>
  filterEntries(tree.nodes, ([nodeId, _]) => isTopLevel(tree.nodes)(nodeId));

/**
 * Receives an array of node IDs in the given tree and returns them sorted by X-coordinate.
 */
export const sortPositionedNodesByXCoord = (tree: PositionedTree) => (nodeIds: Id[]) =>
  Object.entries(associateWith(nodeIds, nodeId => tree.nodes[nodeId].position.treeX))
    .sort(([_, treeX1], [__, treeX2]) => treeX1 - treeX2)
    .map(([nodeId, _]) => nodeId);

/**
 * Returns true if there are no nodes assigned to any slices overlapping the given one.
 */
export const isSliceUnassigned = (tree: PositionedTree) => (slice: StringSlice) => {
  const nodesAssignedToSlice = mapValues(tree.nodes, node => 'slice' in node && slicesOverlap(node.slice, slice));
  return nodesAssignedToSlice.every(overlapsSlice => !overlapsSlice);
};

File: npbloom-core/src/commonMain/kotlin/content/positioned/positionedEntityHelpers.kt
@file:OptIn(ExperimentalJsExport::class)

package content.positioned

import content.Id
import content.IdMap
import content.StringSlice
import content.slicesOverlap

/**
 * Returns nodes matching the given predicate in the given tree.
 */
@JsExport
fun filterPositionedNodesInTree(
    predicate: (node: PositionedNode) -> Boolean,
    tree: PositionedTree,
): IdMap<PositionedNode> = tree.nodes.filterValues(predicate)

/**
 * Returns nodes with the given IDs in the given tree.
 */
@JsExport
fun filterPositionedNodesInTreeById(nodeIds: Set<Id>, tree: PositionedTree): IdMap<PositionedNode> =
    tree.nodes.filterKeys { it in nodeIds }

/**
 * Returns whether the node with the given ID is a top-level node in the given node ID map.
 */
@JsExport
fun isTopLevel(nodes: IdMap<PositionedNode>, nodeId: Id) =
    nodes.none { (_, node) -> node is PositionedBranchingNode && nodeId in node.children }

/**
 * Returns an ID map consisting of the top-level nodes in the given tree.
 */
@JsExport
fun getTopLevelPositionedNodes(tree: PositionedTree): IdMap<PositionedNode> =
    tree.nodes.filterKeys { isTopLevel(tree.nodes, it) }

/**
 * Receives an array of node IDs in the given tree and returns them sorted by X-coordinate.
 */
@JsExport
fun sortPositionedNodesByXCoord(tree: PositionedTree, nodeIds: Set<Id>): Array<Id> =
    nodeIds.sortedBy { tree.nodes[it]!!.position.treeX }.toTypedArray()

/**
 * Returns true if there are no nodes assigned to any slices overlapping the given one.
 */
@JsExport
fun isSliceUnassigned(tree: PositionedTree, slice: StringSlice) =
    tree.nodes.count { (_, node) -> node is PositionedTerminalNode && slicesOverlap(node.slice, slice) } == 0
